### ISSUE TITLE
solaris open file descriptor recursive lib loading error.

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#6640]
Migrated from [rt.perl.org#23042](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=23042) (status was 'resolved')

Searchable as RT23042$

**Comment by p5pRT** [comment]
### From craig@malchiarts.com

#### Created by craig@malchiarts.com



Solaris has a known issue with open file descriptors\, but the error message
thrown when the limit has been reached by loading a lib is misleading\.  Also\,
recursive library calls \(libA calls libB calls libC etc\.\.\.\) generate the same
error message\.

use program below for reference&#8203;:

\#\!/usr/bin/perl \-w

use lib "\.";
use IO&#8203;::File;
use strict;
use Getopt&#8203;::Long;

my @&#8203;files;
my $load = 1;
my $max = 251;

GetOptions\( "load|l=s" => \\$load\,
&nbsp;                "max|m=s" => \\$max
&nbsp;         \);

print "Opening files\.\.\. ";
foreach \(0\.\.$max\) \{
&nbsp;    push @&#8203;files\, IO&#8203;::File\->new\(">aa$\_"\);
\}
print "count&#8203;: " \. scalar\(@&#8203;files\) \."\\n";

if \( $load == 1 \) \{
&nbsp;    print "loading 1 lib\.\\n";
&nbsp;    require "fhtestlib\.pm";
\} elsif \( $load == 2 \) \{
&nbsp;    print "loading 2 libs\.\\n";
&nbsp;    require "fhtestlib\.pm";
&nbsp;    require "fhtestlib2\.pm";
\} elsif \( $load == 3 \) \{
&nbsp;    print "loading 1 lib\, which loads 2 libs\\n";
&nbsp;    require "fhtestlib3\.pm";
\} else \{
&nbsp;    print "invalid load \(\-l\) \\n";

\}


print "Cleaning up\.\.\.";
foreach \(0 \.\. $max\) \{
&nbsp;    $files\[$\_\]\->close\(\);
\}
&nbsp;    print \`rm aa\*\`;
print "\\n";
1;

\#\# put these packages in separate files \#\#
package fhtestlib;


print "fhtestlib loaded\.\\n";

1;

package fhtestlib2;

print "fhtestlib2 loaded\\n";

1;

package fhtestlib3;

use lib "\.";
use fhtestlib;
use fhtestlib2;

print "fhtestlib3 loaded\.\\n";

1;


also note that adding a \-m switch greater than the accepted amount will not
cause an error\. \(ie&#8203;: \-m 400\) until something is attempted to be done with the
file handle\.  \(I don't know if that's an IO&#8203;::File bug or no\.\.\.\)\.

I suppose it's not such a big deal\, but it's come up twice at work\.\.\. first
time took some time to figure it out\.\.\.

\(note&#8203;:  below site configuration is \*not\* where I'm having the problem ;\)

\-craig

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl v5.8.0:

Configured by root at Thu Feb 27 12:12:44 PST 2003.

Summary of my perl5 (revision 5.0 version 8 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.4.20, archname=i386-linux
    uname='linux midas 2.4.20 #4 sat feb 22 18:47:21 pst 2003 i686 unknown '
    config_args='-de -Dprefix=/usr -Dcccdlflags=-fPIC -Darchname=i386-linux'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O3',
    cppflags='-fno-strict-aliasing'
    ccversion='', gccversion='3.2.2', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lgdbm -ldb -ldl -lm -lc -lcrypt -lutil
    perllibs=-lnsl -ldl -lm -lc -lcrypt -lutil
    libc=/lib/libc-2.3.1.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.3.1'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fPIC', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.0:
    /usr/lib/perl5/5.8.0/i386-linux
    /usr/lib/perl5/5.8.0
    /usr/lib/perl5/site_perl/5.8.0/i386-linux
    /usr/lib/perl5/site_perl/5.8.0
    /usr/lib/perl5/site_perl
    .


Environment for perl v5.8.0:
    HOME=/usr/home/craig
    LANG=C
    LANGUAGE (unset)
    LC_ALL=POSIX
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/opt/kde/bin:/usr/local/sbin:/usr/sbin:/sbin:/usr/local/bin:/usr/bin:/bin:/usr/X11R6/bin:/usr/games:/usr/lib/qt-3.1.2/bin:/usr/share/texmf/bin:.
    PERL_BADLANG (unset)
    SHELL=/bin/tcsh

```
</details>

**Comment by p5pRT** [comment]
### From @eserte

"craig@&#8203;malchiarts\.com \(via RT\)" \<perlbug\-followup@&#8203;perl\.org> writes&#8203;:

> \# New Ticket Created by  craig@&#8203;malchiarts\.com 
> \# Please include the string&#8203;:  \[perl \#23042\]
> \# in the subject line of all future correspondence about this issue\. 
> \# \<URL&#8203;: http&#8203;://rt\.perl\.org/rt2/Ticket/Display\.html?id=23042 >
> 
> 
> 
> This is a bug report for perl from craig@&#8203;malchiarts\.com\,
> generated with the help of perlbug 1\.34 running under perl v5\.8\.0\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
> 
> Solaris has a known issue with open file descriptors\,

You can reproduce the problem on other systems\, too\. Just set the file
descriptor limit to a low value \(e\.g\. by using limits \-n\)\. Other
systems like Linux and FreeBSD have much higher default limits than
Solaris\.

> but the error message
> thrown when the limit has been reached by loading a lib is misleading\.  Also\,
> recursive library calls \(libA calls libB calls libC etc\.\.\.\) generate the same
> error message\.
> 
> use program below for reference&#8203;:
> 
\[\.\.\.\]
> 
> 
> also note that adding a \-m switch greater than the accepted amount will not
> cause an error\. \(ie&#8203;: \-m 400\) until something is attempted to be done with the
> file handle\.  \(I don't know if that's an IO&#8203;::File bug or no\.\.\.\)\.
> 
> I suppose it's not such a big deal\, but it's come up twice at work\.\.\. first
> time took some time to figure it out\.\.\.
> 
> \(note&#8203;:  below site configuration is \*not\* where I'm having the problem ;\)
> 
> \-craig
> 
> 

In principle this could be solved by a little more diagnostics\. e\.g\.&#8203;:

<details><summary>Inline Patch</summary>

```diff
--- bleedperl/pp_ctl.c	Wed Oct  8 14:34:00 2003
+++ bleedperl2/pp_ctl.c	Sat Oct 11 15:26:59 2003
@@ -3266,6 +3266,10 @@ PP(pp_require)
 		}
 		sv_catpvn(msg, ")", 1);
 		SvREFCNT_dec(dirmsgsv);
+		if (errno && errno != ENOENT) {
+		    sv_catpvn(msg, ": ", 2);
+		    sv_catpv(msg, Strerror(errno));
+		}
 		msgstr = SvPV_nolen(msg);
 	    }
 	    DIE(aTHX_ "Can't locate %s", msgstr);

-- 
```
</details>

Slaven Rezic \- slaven@&#8203;rezic\.de
&nbsp;  BBBike \- route planner for cyclists in Berlin
&nbsp;  WWW version&#8203;:                           http&#8203;://www\.bbbike\.de
&nbsp;  Perl/Tk version for Unix and Windows&#8203;:  http&#8203;://bbbike\.sourceforge\.net

**Comment by p5pRT** [comment]
### From @smpeters

On Sat Oct 11 06&#8203;:35&#8203;:36 2003\, slaven@&#8203;rezic\.de wrote&#8203;:
> "craig@&#8203;malchiarts\.com \(via RT\)" \<perlbug\-followup@&#8203;perl\.org> writes&#8203;:
> 
> > \# New Ticket Created by  craig@&#8203;malchiarts\.com
> > \# Please include the string&#8203;:  \[perl \#23042\]
> > \# in the subject line of all future correspondence about this issue\.
> > \# \<URL&#8203;: http&#8203;://rt\.perl\.org/rt2/Ticket/Display\.html?id=23042 >
> >
> >
> >
> > This is a bug report for perl from craig@&#8203;malchiarts\.com\,
> > generated with the help of perlbug 1\.34 running under perl v5\.8\.0\.
> >
> >
> > \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> > \[Please enter your report here\]
> >
> > Solaris has a known issue with open file descriptors\,
> 
> You can reproduce the problem on other systems\, too\. Just set the file
> descriptor limit to a low value \(e\.g\. by using limits \-n\)\. Other
> systems like Linux and FreeBSD have much higher default limits than
> Solaris\.
> 
> > but the error message
> > thrown when the limit has been reached by loading a lib is
> misleading\.  Also\,
> > recursive library calls \(libA calls libB calls libC etc\.\.\.\) generate
> the same
> > error message\.
> >
> > use program below for reference&#8203;:
> >
> \[\.\.\.\]
> >
> >
> > also note that adding a \-m switch greater than the accepted amount
> will not
> > cause an error\. \(ie&#8203;: \-m 400\) until something is attempted to be done
> with the
> > file handle\.  \(I don't know if that's an IO&#8203;::File bug or no\.\.\.\)\.
> >
> > I suppose it's not such a big deal\, but it's come up twice at
> work\.\.\. first
> > time took some time to figure it out\.\.\.
> >
> > \(note&#8203;:  below site configuration is \*not\* where I'm having the
> problem ;\)
> >
> > \-craig
> >
> >
> 
> In principle this could be solved by a little more diagnostics\. e\.g\.&#8203;:
> 
> \-\-\- bleedperl/pp\_ctl\.c    Wed Oct  8 14&#8203;:34&#8203;:00 2003
> \+\+\+ bleedperl2/pp\_ctl\.c    Sat Oct 11 15&#8203;:26&#8203;:59 2003
> @&#8203;@&#8203; \-3266\,6 \+3266\,10 @&#8203;@&#8203; PP\(pp\_require\)
>          \}
>          sv\_catpvn\(msg\, "\)"\, 1\);
>          SvREFCNT\_dec\(dirmsgsv\);
> \+        if \(errno && errno \!= ENOENT\) \{
> \+            sv\_catpvn\(msg\, "&#8203;: "\, 2\);
> \+            sv\_catpv\(msg\, Strerror\(errno\)\);
> \+        \}
>          msgstr = SvPV\_nolen\(msg\);
>          \}
>          DIE\(aTHX\_ "Can't locate %s"\, msgstr\);
> 

This problem was resolved as of change \#25677\.

**Comment by p5pRT** [comment]
@smpeters - Status changed from 'open' to 'resolved'
